3.analyze_data/utils/isolation_forest_data_feature_importance.py

```diff
@@ -0,0 +1,148 @@
+from collections import defaultdict
+from typing import Optional, Sequence, Union
+
+import numpy as np
+import pandas as pd
+from joblib import Parallel, delayed
+from sklearn.tree import DecisionTreeRegressor, _tree
+
+
+class IsoforestFeatureImportance:
+    # Computes Isolation Forest (Scikit-learn) feature importances from a subset of data.
+
+    def __init__(
+        self,
+        _estimators: Sequence[DecisionTreeRegressor],
+        _morphology_data: pd.DataFrame,
+    ):
+        """
+        Parameters
+        ----------
+        _estimators: Scikit-learn decision tree regressor estimators.
+        _morphology_data: Data with only numerical morphology data (without metadata).
+        """
+
+        self._estimators = _estimators
+        self._morphology_data = _morphology_data
+        self._isoforest_importances = None
+
+    @property
+    def isoforest_importances(self):
+        if self._isoforest_importances is None:
+            raise ValueError("isoforest_importances have not been computed")
+
+        return self._isoforest_importances
+
+    def save_tree_feature_importances(
+        self,
+        _tree_obj: _tree.Tree,
+        _leaf_id: int,
+        _sample_idx: int,
+    ) -> dict[dict[str, float]]:
+        """
+        Aggregates and then returns feature importances per tree.
+
+        Parameters
+        _tree_obj: Object for traversing tree.
+        _leaf_id: Leaf node for the sample in the tree.
+        _sample_idx: Sample location in the dataframe.
+        """
+
+        node_id = 0  # Start at the root node
+        depth = 0
+        feature_importances = defaultdict(list)
+        morphology_features = self._morphology_data.iloc[_sample_idx].copy()
+
+        while node_id != _leaf_id:
+            feature_idx = _tree_obj.feature[node_id]
+
+            if feature_idx >= 0:  # Ignore leaf nodes (-2)
+                """
+                Samples are often easier to seperate with less data.
+                Therefore, earlier features have greater importances.
+                """
+                feature_importances[self._morphology_data.columns[feature_idx]].append(
+                    1 / (1 + depth)
+                )
+
+            if morphology_features.iloc[feature_idx] <= _tree_obj.threshold[node_id]:
+                node_id = _tree_obj.children_left[node_id]
+            else:
+                node_id = _tree_obj.children_right[node_id]
+
+            depth += 1
+
+        """
+        Gives a balanced view of expected feature importance in the tree.
+        As opposed to the total contribution, which increases with the (same) features.
+        Since features are choosen randomly, dividing feature importances
+        by the depth weights each feature by how soon the data could be isolated.
+        """
```

Comment:
An extra bit of clarity here would be helpful

1. What do you mean by "balanced view"?
2. When you say "how soon the data could be isolated", do you mean "how soon the feature is able to isolate a single sample in the tree"?